contracts/staking/PoSValidatorManager.sol

```diff
@@ -129,25 +139,44 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
         uint32 messageIndex
     ) external {
         PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        Validator memory validator = _initializeEndValidation(validationID);
+
         // Check that minimum stake duration has passed
-        Validator memory validator = getValidator(validationID);
         require(
-            block.timestamp
+            validator.endedAt
                 >= validator.startedAt + $._validatorRequirements[validationID].minStakeDuration,
             "PoSValidatorManager: minimum stake duration not met"
         );
 
         if (includeUptimeProof) {
-            _getUptime(validationID, messageIndex);
+            uint64 uptime = _getUptime(validationID, messageIndex);
+            if (
+                uptime * 100
+                    >= (validator.endedAt - validator.startedAt) * UPTIME_REWARDS_THRESHOLD_PERCENTAGE
+            ) {
+                $._redeemableValidatorRewards[validator.owner] += $
+                    ._rewardCalculator
+                    .calculateReward(
+                    weightToValue(validator.startingWeight),
+                    validator.startedAt,
+                    validator.endedAt,
+                    0,
+                    0
+                );
+            }
         }
-        // TODO: Calculate the reward for the validator, but do not unlock it
-
-        _initializeEndValidation(validationID);
     }
 
     function completeEndValidation(uint32 messageIndex) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
         Validator memory validator = _completeEndValidation(messageIndex);
-        _unlock(validator.startingWeight, validator.owner);
+
+        _reward(validator.owner, $._redeemableValidatorRewards[validator.owner]);
```

Comment:
this should be check to be PoS Validator, if it's a PoA after migration just does nothign else. Also calling since these functions make external calls for `token.safeTransfer` or mint, we need to watch out for reentrancy. I should have it added already in this PR though https://github.com/ava-labs/teleporter/pull/547/files. 

Can you change the ordering to adhere to check-effects-interaction pattern? Would cause an extra read of reading the state variable redeemable rewards to a local variable, deleting state, then making the external calls, but think it's worth for security and readability.

Reply:
So, I tested out adding a check for `validator.owner == address(0)` here, but without your other changes to move the `owner` to be PoS-only, it doesn't currently do anything for the migration case.